## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-zoned.sh | :x: [more...](#a13bf259-c5e6-40cb-991d-a218635f7aee) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#641541a9-3220-44a9-b36c-443c86dd6b56) |
<a name="641541a9-3220-44a9-b36c-443c86dd6b56"></a>
### 12190-d9a185b/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30230
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 234a9aea990e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2633,
  "function": "aws --endpoint-url http://minio-dev8.minio.io:30230 s3api copy-object --bucket awscli-mint-test-bucket-10343 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-10343/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 46 seconds
(7/15) Running minio-dotnet tests ... done in 47 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 12 seconds
{
  "args": {},
  "duration": 798,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 12 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 131,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-2s49hkd, objectName = minio-java-test-3n3ijvu-copy, resource = /minio-java-test-2s49hkd/minio-java-test-3n3ijvu-copy, requestId = 167A641583FBD102, hostId = 6a882eaa-562d-4e4b-97b8-d254398629cc)\nrequest={method=PUT, url=http://minio-dev8.minio.io:30230/minio-java-test-2s49hkd/minio-java-test-3n3ijvu-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-1ab2imh/minio-java-test-3n3ijvu\nHost: minio-dev8.minio.io:30230\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210429T172419Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210429/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Thu, 29 Apr 2021 17:24:19 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 167A641583FBD102\nX-Xss-Protection: 1; mode=block\nDate: Thu, 29 Apr 2021 17:24:19 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 48 seconds
(9/15) Running minio-py tests ... done in 2 minutes and 58 seconds
(10/15) Running s3cmd tests ... FAILED in 6 seconds
{
  "name": "s3cmd",
  "duration": "3408",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-5601/s3cmd-test-object-14977 56531662-3fdb-4a53-b1f9-089952c61ee5\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 56531662-3fdb-4a53-b1f9-089952c61ee5 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
<a name="a13bf259-c5e6-40cb-991d-a218635f7aee"></a>
### 12190-d9a185b/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30228
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c33b5588c3d0:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... done in 1 minutes and 22 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 36 seconds
(8/15) Running minio-dotnet tests ... done in 43 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 16 seconds
(10/15) Running minio-java tests ... done in 2 minutes and 5 seconds
(11/15) Running minio-js tests ... done in 48 seconds
(12/15) Running minio-py tests ... FAILED in 2 minutes and 37 seconds
{
  "name": "minio-py:test_thread_safe",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-4cb0ea4e-6ef7-4fd6-86b6-d38a6e91e278",
    "object_name": "08b1c858-9137-4738-9315-886f83b7e414"
  },
  "message": "S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-4cb0ea4e-6ef7-4fd6-86b6-d38a6e91e278, request_id: 167A644D5FC5CA8E, host_id: f1edcb6f-ef9a-4c3f-8374-63b429641843, bucket_name: minio-py-test-4cb0ea4e-6ef7-4fd6-86b6-d38a6e91e278",
  "error": "Traceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 1640, in test_thread_safe\n    _CLIENT.remove_bucket(bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 676, in remove_bucket\n    self._execute(\"DELETE\", bucket_name)\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 409, in _execute\n    preload_content=preload_content,\n  File \"/usr/local/lib/python3.6/dist-packages/minio/api.py\", line 385, in _url_open\n    raise response_error\nminio.error.S3Error: S3 operation failed; code: BucketNotEmpty, message: The bucket you tried to delete is not empty, resource: /minio-py-test-4cb0ea4e-6ef7-4fd6-86b6-d38a6e91e278, request_id: 167A644D5FC5CA8E, host_id: f1edcb6f-ef9a-4c3f-8374-63b429641843, bucket_name: minio-py-test-4cb0ea4e-6ef7-4fd6-86b6-d38a6e91e278\n",
  "duration": 6825
}
(12/15) Running s3cmd tests ... done in 20 seconds
(13/15) Running s3select tests ... done in 6 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/cmd/config-encrypted.go
+++ b/cmd/config-encrypted.go
@@ -94,8 +94,6 @@ func migrateIAMConfigsEtcdToEncrypted(ctx context.Context, client *etcd.Client) 
 				return err
 			}
 			logger.Info("Attempting to re-encrypt config, IAM users and policies on MinIO with %q (%s)", stat.DefaultKey, stat.Name)
-		} else {
-			logger.Info("Attempting to migrate encrypted config, IAM users and policies on MinIO to a plaintext format. To encrypt all MinIO config data a KMS is needed")
 		}
 	}
 
@@ -139,9 +137,7 @@ func migrateIAMConfigsEtcdToEncrypted(ctx context.Context, client *etcd.Client) 
 
 	if encrypted {
 		if GlobalKMS != nil {
-			logger.Info("Migration of encrypted config data completed. All config data is now encrypted with the KMS")
-		} else {
-			logger.Info("Migration of encrypted config data completed. All config data is now stored in plaintext")
+			logger.Info("Migration of encrypted config data completed. All config data is now encrypted.")
 		}
 	}
 	return deleteKeyEtcd(ctx, client, backendEncryptedFile)
@@ -158,8 +154,6 @@ func migrateConfigPrefixToEncrypted(objAPI ObjectLayer, encrypted bool) error {
 				return err
 			}
 			logger.Info("Attempting to re-encrypt config, IAM users and policies on MinIO with %q (%s)", stat.DefaultKey, stat.Name)
-		} else {
-			logger.Info("Attempting to migrate encrypted config, IAM users and policies on MinIO to a plaintext format. To encrypt all MinIO config data a KMS is needed")
 		}
 	}
 
@@ -201,9 +195,7 @@ func migrateConfigPrefixToEncrypted(objAPI ObjectLayer, encrypted bool) error {
 	}
 	if encrypted {
 		if GlobalKMS != nil {
-			logger.Info("Migration of encrypted config data completed. All config data is now encrypted with the KMS")
-		} else {
-			logger.Info("Migration of encrypted config data completed. All config data is now stored in plaintext")
+			logger.Info("Migration of encrypted config data completed. All config data is now encrypted.")
 		}
 	}
 	return deleteConfig(GlobalContext, globalObjectAPI, backendEncryptedFile)

--- a/docs/kms/IAM.md
+++ b/docs/kms/IAM.md
@@ -8,28 +8,75 @@ onwards, MinIO will use the KMS provided keys to encrypt the IAM data instead
 of the cluster root credentials. If the KMS is not enabled, MinIO will store
 the IAM data as plain text in its backend.
 
+### MinIO KMS Quick Start
+
+MinIO supports two ways of encrypting IAM and configuration data.
+You can either use KES - together with an external KMS - or, much simpler,
+set the env. variable `MINIO_KMS_SECRET_KEY` and start/restart the MinIO server. For more details about KES and how
+to set it up refer to our [KMS Guide](https://github.com/minio/minio/blob/master/docs/kms/README.md).
+
+Instead of configuring an external KMS you can start with a single key by
+setting the env. variable `MINIO_KMS_SECRET_KEY`. It expects the following
+format:
+```sh
+MINIO_KMS_SECRET_KEY=<key-name>:<base64-value>
+```
+
+First generate a 256 bit random key via:
+```sh
+$ cat /dev/urandom | head -c 32 | base64 -
+OSMM+vkKUTCvQs9YL/CVMIMt43HFhkUpqJxTmGl6rYw=
+```
+
+Now, you can set `MINIO_KMS_SECRET_KEY` like this:
+```sh
+export MINIO_KMS_SECRET_KEY=my-minio-key:OSMM+vkKUTCvQs9YL/CVMIMt43HFhkUpqJxTmGl6rYw=
+```
+> You can choose an arbitrary name for the key - instead of `my-minio-key`.
+> Please note that loosing the `MINIO_KMS_SECRET_KEY` will cause data loss
+> since you will not be able to decrypt the IAM/configuration data anymore.
+For distributed MinIO deployments, specify the *same* `MINIO_KMS_SECRET_KEY` for each MinIO server process. 
+
+At any point in time you can switch from `MINIO_KMS_SECRET_KEY` to a full KMS
+deployment. You just need to import the generated key into KES - for example via
+the KES CLI once you have successfully setup KES:
+```sh
+kes key create my-minio-key OSMM+vkKUTCvQs9YL/CVMIMt43HFhkUpqJxTmGl6rYw=
+```
+- For instructions on setting up KES, see the [KES Getting Started guide](https://github.com/minio/kes/wiki/Getting-Started) 
+
+- For instructions on using KES for encrypting the MinIO backend, follow the [KMS Quick Start](https://github.com/minio/minio/tree/master/docs/kms). The SSE-S3 configuration setup also supports MinIO KMS backend encryption.
+
 ### FAQ
 
-> Why is this change needed? Was the previous encryption not secure (enough)?
+> Why is this change needed?
 
-The previous mechanism of encryption using the cluster root credentials itself
-is secure. The purpose of this change is to improve the encryption key
-management by delegating it to the KMS. Once the cluster root credentials are no
-longer needed to decrypt persistent data at the backend, they can be rotated
-much more easily. By using a KMS to protect the IAM data a potential security or
-compliance team has much more control over how/when a MinIO cluster can be
-started.
+Before, there were two separate mechanisms - S3 objects got encrypted using a KMS,
+if present, and the IAM / configuration data got encrypted with the root credentials.
+Now, MinIO encrypts IAM / configuration and S3 objects with a KMS, if present. This
+change unified the key-management aspect within MinIO.
+
+The unified KMS-based approach has several advantages:
+ - Key management is now centralized. There is one way to change or rotate encryption keys.
+   There used to be two different mechanisms - one for regular S3 objects and one for IAM data.
+ - Reduced server startup time. For IAM encryption with the root credentials, MinIO had
+   to use a memory-hard function (Argon2) that (on purpose) consumes a lot of memory and CPU.
+   The new KMS-based approach can use a key derivation function that is orders of magnitudes
+   cheaper w.r.t. memory and CPU.
+ - Root credentials can now be changed easily. Before, a two-step process was required to
+   change the cluster root credentials since they were used to en/decrypt the IAM data.
+   So, both - the old and new credentials - had to be present at the same time during a rotation
+   and the old credentials had to be removed once the rotation completed. This process is now gone.
+   The root credentials can now be changed easily.
 
 > Does this mean I need an enterprise KMS setup to run MinIO (securely)?
 
 No, MinIO does not depend on any third-party KMS provider. You have three options here:
 
 - Run MinIO without a KMS. In this case all IAM data will be stored in plain-text.
-
 - Run MinIO with a single secret key. MinIO supports a static cryptographic key
   that can act as minimal KMS. With this method all IAM data will be stored
   encrypted. The encryption key has to be passed as environment variable.
-
 - Run MinIO with KES (minio/kes) in combination with any supported KMS as
   secure key store. For example, you can run MinIO + KES + Hashicorp Vault.
 

